### ISSUE TITLE
clarification on MAs_t

### ISSUE THREAD

**Issue by BenjaminBeichler** [openETCS/modeling#154]
@JakobGartner 

To extract the SvL I need to know whether DP or OL is set. Since there are no flags currently, I take the Point which have the greatest Position of the 3 values DangerPoint,Overlap,EOA. 

So If DangerPoint or Overlap is not valid in MA then corresponding model should set the value to the position of the EOA. If you intend to do it in a nother way, please let me know.

**Comment by JakobGartner** [comment]
up to now I have a logic in mind that follows the logic of the packets where there is a field that indicates whether there is a DP or OL. If you need the flag it is easy for me to generate it in the output function

**Comment by BenjaminBeichler** [comment]
then please generate the flag and update the datatype (sorry for make you so busy )

**Comment by JakobGartner** [comment]
OK I will do.

Do you want it in the data structure or separately ?
Do you need a flag “DP or OL” or do you need a flag per location separately.

Remark:

In our test cases, this flags will always stay “false” because we don’t have this kind of location in our test data

**Comment by BenjaminBeichler** [comment]
as in §3.8.4.5 of the SRS, I need the flag for every of the points

**Comment by JakobGartner** [comment]
ok you will get it. 


I will think about a good way to implement it

(yes it is in my internal data already)

**Comment by JakobGartner** [comment]
committed (done)

**Comment by BenjaminBeichler** [comment]
@JakobGartner 

unfortunataly the type definition is not complete (_null insteat of bool)

**Comment by JakobGartner** [comment]
it seems that that was a path problem (subtypes painting to standard oEZTS libs in wrong path)

I hope it's fixed now

**Comment by BenjaminBeichler** [comment]
@JakobGartner 

there is another Problem ... the ReleaseSpeed of the MA needs also a valid flag, since it could be defined at several places and we need to choose it somehow. I propose for the review not to change the type now again, but later. I will choose now the most restrictive release speed, although it's not important for our usecases.

**Comment by BenjaminBeichler** [comment]
@JakobGartner  @BerndHekele 

not only the releasespeed need this flag, also the whole array item need the valid flag, which was in before the versions were mixed up. To deliver today the working model I need this more or less immediately

**Comment by JakobGartner** [comment]
I just proposed an update. @BenjaminBeichler: Please check and if it's not what you need let's have a direct call to sync it once and for all.

Added valid flag per MA section
Added valid flag as request for Release Speed.

**Comment by T12z** [comment]
Hej @JakobGartner ,
as I just tried to finish the onboard calculation of the release speed, I am not quite satisfied with the inputs I get. In the spec that defines the trackside values for the release speed (7.5.1.168..169), it
 1. defines a special resolution,
 2. defines special / magic values.

Who does the conversion, what resolution/unit is passed to SpeedDistanceMon? How are the special values passed (I actually need to do decisions on them)? I could work with the raw value but this might not be apropriate on an architectural approach. The value flag discussed earlier should still validate the existance of the release speed.

And on a broader discussion, any idea how to handle the two definitions (168 and 169) differently and if both are given? I am not too confident to have resembled a complete understanding through the spec.
cheers, Thorsten

**Comment by JakobGartner** [comment]
The Operators TM_conversions::CAST_V_RELEASEDP_to_int and M_conversions::CAST_V_RELEASEOL_to_int handle the conversion transparently:

the 7bit value from trackside is converted to an integer value 0<= V.. <= 600km/h

(the resolution is in 5 km/h steps but you don’t need to care unless it is relevant for your numeric model)

This transformation operator NOT handling the special values.


I am also working on the SP and MA management functions and was thinking about doing the same as for the MA timers. Handling those things transparently (you get profiles where these switches have already been evaluated) but I am equally not si sure about what your function does in detail with them.

We should have a call 

Jakob

**Comment by T12z** [comment]
Hej @JakobGartner ,

this issue seems to have received a low priority tag. I will anyways propose a no-frills solution:

the way you send out the release speed at the moment (v-value and valid-tag) should be extended to
 * keep the release speed as given by the MA in a standard format (I guess that is in integer [km/h])
 * an enumeration representing {"not valid", "given by MA", "use NV", "calculate onboard"} -- though the naming should rather reflect the naming from the SRS.

"not valid" means the package / record has not been seen yet, as it should always exist by standard's definition.

I still have no idea how to handle the two different types of Release speed. Are they mutually exclusive? My operator only receives one and my part of the SRS only knows about ***a*** release speed.

**Comment by mairamou** [comment]
Well, I have many high- priority tags yes, this was not the first one (the ones in front of the data flow impact more people)

I have now finished all the stuff on the TrackMessages module, and we are still testing it, but I will now go to the heart of the system: The TrackAtlas, and this includes this topic.

I will be able to look at your question today. 

Jakob

**Comment by mairamou** [comment]
I am a bit confused because Github seems to send my mail as if it doesn’t come from myself but from my colleague

**Comment by JakobGartner** [comment]
When replying to issues by email, Github marks my messages as if they were from Mairamou.
Maybe we have a problem with mail setup, I will check.

**Comment by JakobGartner** [comment]
Well, I have many high- priority tags yes, this was not the first one (the ones in front of the data flow impact more people)

I have now finished all the stuff on the TrackMessages module, and we are still testing it, but I will now go to the heart of the system: The TrackAtlas, and this includes this topic.

I will be able to look at your question today. 

Jakob

**Comment by JakobGartner** [comment]
I have looked at it.

1. The TrackMessages library does some filtering
2. The TrackAtlas library will process the special flags and hand them to you.
3. The SRS does not directly provide variable types for each of these scenarios, but I will use the original SRS wording. 
4. The TrackMessages library already converts the variables that have a fixed conversion factor (like V and T types)
5. The TrackAtlas will convert D types according to Q_SCALE.

**Comment by JakobGartner** [comment]
by the way, my emails were tagged by github the wrong way because I had replied to emails that Mairamou had forwarded to me. Apparently Github is using a token and not the sender address to identify the sender. Sorry for the confusion.

**Comment by JakobGartner** [comment]
I will close this work end of the week.

**Comment by BaseliyosJacob** [comment]
@BenjaminBeichler is this issue solved?

**Comment by BenjaminBeichler** [comment]
@T12z  is waiting on the solution of his question about the special values of release speeds. I can not assure that this is already solved by @JakobGartner.

All other questions are solved.

**Comment by JakobGartner** [comment]
after further analysis here some explanation how the MAs was supposed to work- it was one of these "overloaded" data types following the examples found in the model when I started work.

The information found in the structure is to be understood following these principles:

A "section" can be one of the following:

1- A normal MA section
2- An end section
3- An OL
4- A DP
5- A LOA

Following this concept, the information found in each section will contain the subset of information to characterise this section. 

The flags EndSection, DP_valid, OL_valid etc. are mutually exclusive.
This means that if DP_valid = true, this implies that if there is a ReleaseSpeed defined, this refers either to OL or DP depending on the flag.

After my current analysis I believe that this "overloading" should not be stressed so much and I propose a data structure which is closer to the SRS.

In order to minimise impact, I propose the following (comments in the screenshot)
![screen shot 2015-07-14 at 13 24 50](https://cloud.githubusercontent.com/assets/8400941/8672135/c2198710-2a2b-11e5-9b01-b5550443fa45.png)

![screen shot 2015-07-14 at 13 27 45](https://cloud.githubusercontent.com/assets/8400941/8672174/2c8bf9d4-2a2c-11e5-901f-bc6949ef2cd0.png)

![screen shot 2015-07-14 at 13 30 20](https://cloud.githubusercontent.com/assets/8400941/8672210/8349f5be-2a2c-11e5-8516-133d6a34f56e.png)
![screen shot 2015-07-14 at 13 32 20]
(https://cloud.githubusercontent.com/assets/8400941/8672238/d27a823e-2a2c-11e5-8151-14025550c7b8.png)

**Comment by JakobGartner** [comment]
All speeds/ distance are given in the OBU basic types L/T/V.
All special/magic values are already resolved (for example, the NV speed is given if this was defined by the special value)

**Comment by T12z** [comment]
MA_t does not exist anymore. Being stale and nobody kehrs, I will swipe this off the list.
